controller/restore_controller.go

```diff
@@ -0,0 +1,210 @@
+		DeleteFunc: rsc.enqueueVolume,
+	})
+	rsc.cacheSyncs = append(rsc.cacheSyncs, ds.VolumeInformer.HasSynced)
+
+	return rsc
+}
+
+func (rsc *RestoreController) enqueueVolume(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	rsc.queue.Add(key)
+}
+
+func (rsc *RestoreController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer rsc.queue.ShutDown()
+
+	rsc.logger.Infof("Start Longhorn restore controller")
+	defer rsc.logger.Infof("Shutting down Longhorn restore controller")
+
+	if !cache.WaitForNamedCacheSync(rsc.name, stopCh, rsc.cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(rsc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (rsc *RestoreController) worker() {
+	for rsc.processNextWorkItem() {
+	}
+}
+
+func (rsc *RestoreController) processNextWorkItem() bool {
+	key, quit := rsc.queue.Get()
+	if quit {
+		return false
+	}
+	defer rsc.queue.Done(key)
+	err := rsc.syncHandler(key.(string))
+	rsc.handleErr(err, key)
+	return true
+}
+
+func (rsc *RestoreController) handleErr(err error, key interface{}) {
+	if err == nil {
+		rsc.queue.Forget(key)
+		return
+	}
+
+	rsc.logger.WithError(err).Warnf("Error syncing Longhorn volume %v", key)
+	rsc.queue.AddRateLimited(key)
+	return
+}
+
+func (rsc *RestoreController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: failed to sync volume %v", rsc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != rsc.namespace {
+		return nil
+	}
+	return rsc.reconcile(name)
+}
+
+func (rsc *RestoreController) reconcile(volName string) (err error) {
+	vol, err := rsc.ds.GetVolume(volName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	if !rsc.isResponsibleFor(vol) {
+		return nil
+	}
+
+	if vol.Status.RestoreRequired {
+		// TODO: need to create a func to convert volumeName to vaName
+		va, err := rsc.ds.GetLHVolumeAttachment(volName)
+		if err != nil {
+			return err
+		}
+
+		if va.Spec.Attachments == nil {
+			va.Spec.Attachments = make(map[string]*longhorn.Attachment)
+		}
+
+		var restoringAttachment *longhorn.Attachment
+		for _, attachment := range va.Spec.Attachments {
+			if attachment.ID == "restore-controller-"+volName {
+				restoringAttachment = attachment
+				break
+			}
+		}
+
+		if restoringAttachment == nil {
+			//create new one
+			restoringAttachment = &longhorn.Attachment{
+				ID:     "restore-controller-" + volName,
+				Type:   longhorn.AttacherTypeRestoreController,
+				NodeID: vol.Status.OwnerID,
+				Parameters: map[string]string{
+					"disableFrontend": "true",
+				},
+			}
+		}
+
+		if restoringAttachment.NodeID != vol.Status.OwnerID {
+			restoringAttachment.NodeID = vol.Status.OwnerID
+		}
```

Comment:
Any concern about this switch if there is already one that exists?

Reply:
When vol.Status.OwnerID switched to a new value, it means the current  node is down. So we switch the restoringAttachment to new node directly. AD controller should be able to handle this transition smoothly